src/libraries/System.Net.Http/src/System/Net/Http/HttpMetrics.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+
+namespace System.Net.Http
+{
+    internal sealed class HttpMetrics
+    {
+        private readonly Meter _meter;
+        private readonly UpDownCounter<long> _currentRequests;
+        private readonly Histogram<double> _requestsDuration;
+
+        public HttpMetrics(Meter meter)
+        {
+            _meter = meter;
+
+            _currentRequests = _meter.CreateUpDownCounter<long>(
+                "current-requests",
+                description: "Number of outbound HTTP requests that are currently active on the client.");
+
+            _requestsDuration = _meter.CreateHistogram<double>(
+                "request-duration",
+                unit: "s",
+                description: "The duration of outbound HTTP requests.");
+        }
+
+        public void RequestStart(HttpRequestMessage request)
+        {
+#pragma warning disable SA1129 // Do not use default value type constructor
+            var tags = new TagList();
+#pragma warning restore SA1129 // Do not use default value type constructor
+            InitializeCommonTags(ref tags, request);
+            _currentRequests.Add(1, tags);
+        }
+
+        public void RequestStop(HttpRequestMessage request, HttpResponseMessage? response, Exception? unhandledException, long startTimestamp, long currentTimestamp)
+        {
+#pragma warning disable SA1129 // Do not use default value type constructor
+            var tags = new TagList();
+#pragma warning restore SA1129 // Do not use default value type constructor
+            InitializeCommonTags(ref tags, request);
```

Comment:
Because of redirects you may log different tags here than at `RequestStart`.

Reply:
Maybe we could define a type (`struct`?) to encapsulate the metrics context of an outgoing request, a

Reply:
Properties on `HttpRequestMessage` are updated with a response redirect?

In that case, what request values should be used?

Options:
1. Don't use the initial request values anywhere. That means we don't include tags on `current-requests` and have a dimensionless counter. The `request-duration` counter uses the `HttpRequestMessage` state based on when `SendAsync` returns.
2. Use initial request values with `current-requests` and `request-duration`. Ignore request message state after `SendAsync` returns. Maybe have an extra counter to check if the URL changes between sending the request and receiving the response to record redirect counts.
3. A mix. Initial values on `current-requests` and potentially updated values on `request-duration`.

No perfect choice. I like option 2, but curious what client networking experts think.

Reply:
> Properties on `HttpRequestMessage` are updated with a response redirect?

Yes, the request is being reused then to save allocations, but I think this is a fundamental problem, regardless of our implementation.

>  what current diagnostics does.

I don't think this concern exists in case of [existing counters](https://learn.microsoft.com/en-us/dotnet/core/diagnostics/available-counters#systemnethttp-counters). The only ones where we have something similar to tags are the `<protocol>-connections-current-total` and the `<protocol>-requests-queue-duration` counters. In both cases the protocol version is clear at the moment we log counter changes.

> Use initial request values with `current-requests` and `request-duration`. Ignore request message state after `SendAsync` returns.

Wouldn't this be unexpected to many users? For example by using `HttpVersionPolicy.RequestVersionOrLower/RequestVersionOrHigher` and negotiating lower/higher versions, a user would not log the actual HTTP versions being used.

Isn't it also an option to record multiple requests on redirects? (Even though it means that the implementation would be bound to `SocketsHttpHandler` and other handlers would need to implement the metrics separately.)

I have opened https://github.com/open-telemetry/opentelemetry-specification/discussions/3462, hoping that OTel folks can help clarifying this.

Reply:
> I don't think this concern exists in case of [existing counters](https://learn.microsoft.com/en-us/dotnet/core/diagnostics/available-counters#systemnethttp-counters).

Yeah existing counters don't have request-specific info on them, but I was thinking about what is in logs and the activity.

Reply:
I agree with Anton that relying on the request's request version can be quite misleading.
YARP for example uses a default version of 2.0 with a `RequestVersionOrLower` policy.
If we only use the request version, the user would only see 2.0 counters even if they're actually only talking 1.1 to the server.

The unfortunate thing is that we'll only know what version is actually used when we try sending the request to the wire internally, or from a higher-level perspective, once you get back the response message object.

It may make sense to consider moving this logic further down into `SocketsHttpHandler`, especially if you will also want to track when the response content has been read completely.

Reply:
Because the version isn't known when a request starts, it wouldn't be included as a tag on `current-requests`. And `request-duration` would use `HttpResponseMessage.Version` for this information.